vault/diagnose/output.go

```diff
@@ -0,0 +1,174 @@
+package diagnose
+
+import (
+	"context"
+	"io"
+	"strings"
+
+	"go.opentelemetry.io/otel/codes"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	status_unknown = "[      ] "
+	status_ok      = "\u001b[32m[  ok  ]\u001b[0m "
+	status_failed  = "\u001b[31m[failed]\u001b[0m "
+	status_warn    = "\u001b[33m[ warn ]\u001b[0m "
+	same_line      = "\u001b[F"
+	errorStatus    = "error"
+	warnStatus     = "warn"
+	okStatus       = "ok"
+)
+
+type Result struct {
+	Name     string
+	Warnings []string
+	Status   string
+	Message  string
+	Children []*Result
+}
+
+// TelemetryCollector is an otel SpanProcessor that gathers spans and once the outermost
+// span ends, walks the otel traces in order to produce a top-down tree of Diagnose results.
+type TelemetryCollector struct {
+	spans      map[trace.SpanID]sdktrace.ReadOnlySpan
+	rootSpan   sdktrace.ReadOnlySpan
+	results    map[trace.SpanID]*Result
+	RootResult *Result
+}
+
+func NewTelemetryCollector() *TelemetryCollector {
+	return &TelemetryCollector{
+		spans:   make(map[trace.SpanID]sdktrace.ReadOnlySpan),
+		results: make(map[trace.SpanID]*Result),
+	}
+}
+
+func (t *TelemetryCollector) OnStart(parent context.Context, s sdktrace.ReadWriteSpan) {
+	t.spans[s.SpanContext().SpanID()] = s
+
+}
+
+func (t *TelemetryCollector) OnEnd(e sdktrace.ReadOnlySpan) {
+	if !e.Parent().HasSpanID() {
+		// Deep first walk the span structs to construct the top down tree results we want
+		for _, s := range t.spans {
+			r := t.getOrBuildResult(s.SpanContext().SpanID())
+			if r != nil {
+				if s.Parent().HasSpanID() {
+					p := t.getOrBuildResult(s.Parent().SpanID())
+					if p != nil {
+						p.Children = append(p.Children, r)
+					}
+				} else {
+					t.RootResult = r
+				}
+			}
+		}
+	}
+}
+
+func (t *TelemetryCollector) Shutdown(ctx context.Context) error {
```

Comment:
Could you leave some quick comments on these stub methods detailing what we'd want them to do?